File: wmeissner-jruby-cext/example/MathBench/MathBench.c
#include <stdio.h>
#include <ruby.h>

VALUE BenchClass = Qnil;

static VALUE addi2(VALUE self, VALUE i1, VALUE i2);
static VALUE addf2(VALUE self, VALUE f1, VALUE f2);


void
Init_mathbench()
{
    BenchClass = rb_define_class("MathBench", rb_cObject);
    rb_define_method(BenchClass, "addi2", addi2, 2);
    rb_define_method(BenchClass, "addf2", addf2, 2);
}

VALUE
addi2(VALUE self, VALUE i1, VALUE i2)
{
    return INT2NUM(NUM2INT(i1) + NUM2INT(i2));
}

VALUE
addf2(VALUE self, VALUE f1, VALUE f2)
{
    return rb_float_new(NUM2DBL(f1) + NUM2DBL(f2));
}

int 
ffi_addi2(int i1, int i2)
{
  return i1 + i2;
}

double 
ffi_addf2(double f1, double f2)
{
  return f1 + f2;
}

File: wmeissner-jruby-cext/example/MathBench/MathBench.rb
# 
# To change this template, choose Tools | Templates
# and open the template in the editor.
 

$LOAD_PATH.unshift(File.expand_path(File.join(__FILE__, "..")))
require 'mathbench.so'

b = MathBench.new
puts "MathBench.new returned #{b.inspect}"
puts "h.addi2(1, 2) returns #{b.addi2(1, 2)}"
puts "h.addf2(1.1, 2.2) returns #{b.addf2(1.1, 2.2)}"

def ruby_add2(i1, i2)
 i1 + i2
end

require 'ffi'
module Foreign
  extend FFI::Library
  ffi_lib FFI::Library::CURRENT_PROCESS
  attach_function :addi2, :ffi_addi2, [ :int, :int ], :int
  attach_function :addf2, :ffi_addf2, [ :double, :double ], :double
end

puts "Foreign.addi2(1, 2) returns #{Foreign.addi2(1, 2)}"
puts "Foreign.addf2(1.1, 2.2) returns #{Foreign.addf2(1.1, 2.2)}"

require 'benchmark'

iter = 1000_000
10.times do
 Benchmark.bmbm do |bm|
   bm.report("addi2 C (sml)") { iter.times { b.addi2(1, 2) } }
   bm.report("addi2 C (med)") { iter.times { b.addi2(1111, 2222) } }
   bm.report("addf2 C") { iter.times { b.addf2(1.1, 2.2) } }
   bm.report("addi2 FFI") { iter.times { Foreign.addi2(1, 2) } }
   bm.report("addf2 FFI") { iter.times { Foreign.addi2(1.1, 2.2) } }
   bm.report("addi2 Ruby") { iter.times { ruby_add2(1, 2) } }
   bm.report("addf2 Ruby") { iter.times { ruby_add2(1.1, 2.2) } }
 end
end
